packages/optimizer/source.ts

```diff
@@ -105,95 +36,28 @@ export interface Outputs<T> {
   secondary: T[];
 }
 
-const makeImports = () => ({
-  [importModule]: {
-    [importMemoryName]: optimizer.memory,
-    ...Object.fromEntries(
-      [...builtinsTyped.keys()].map((name, i) => [
-        i.toString(36),
-        optimizer[name],
-      ])
-    ),
-  },
-});
-
 /**
- * An instantiated WebAssembly function that can be used either to directly
- * compute outputs and gradients or to step a state via the optimizer.
- *
- * Each WebAssembly module used to construct an instance of this class must
- * follow some conventions using other things exported from this package:
- *
- * - The module name of every import must match `importModule`.
- *
- * - One import must be a memory whose name matches `importMemoryName`.
- *
- * - The rest of the imports must be functions in the same order as the the
- *   insertion order of `builtins`. The name of each of these imports must be a
- *   base-36 integer (`"0"`-`"9"`, `"a"`-`"z"`) corresponding to the index of
- *   the name in `builtins`. The type of the import is determined by its value
- *   in `builtins`, which is an element of `BuiltinType`:
- *
- *   - `"addToStackPointer"` takes one `i32` and returns one `i32`.
- *   - `"unary"` takes one `f64` and returns one `f64`.
- *   - `"binary"` takes two `f64`s and returns one `f64`.
- *   - `"polyRoots"` takes two `i32`s and returns nothing.
- *
- * - The module must export a function whose name matches `exportFunctionName`,
- *   which takes four `i32`s and returns one `f64`.
+ * Computes the sum of several functions on `inputs`, where the corresponding
+ * element of `mask` determines whether that function contributes to the sum.
  */
+export type GradFn = (
+  inputs: Float64Array,
+  mask: Int32Array,
+  gradient: Float64Array,
+  secondary: Float64Array
+) => number;
+
 export class Gradient {
-  private f: WebAssembly.ExportValue;
-  private numAddends: number;
-  private numSecondary: number;
+  f: GradFn;
```

Comment:
In our particular case, this `GradFn` is not actually the exported function from our generated Wasm module, but a JavaScript function that wraps around it so at the boundary we can think in terms of arrays instead of directly in terms of pointers.

Reply:
Gotcha. So really `makeMeta` figures out where the pointers should point to, and `makeCompiled` translates from array args to pointers based on the offsets computed by `makeMeta`.

Reply:
Yeah exactly, so that all happens in the autodiff module, so `@penrose/optimizer` doesn't need to think about pointers at all.